packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMappingRouter.tsx

```diff
@@ -0,0 +1,126 @@
+
+export const GroupingMappingRouter = ({
+  currentRoute,
+  navigateTo,
+  goBack,
+}: {
+  currentRoute: Route;
+  navigateTo: (toRoute: (prev: Route | undefined) => Route) => void;
+  goBack: () => void;
+}) => {
+  const { iModelId } = useGroupingMappingApiConfig();
+  const { mapping, group, groupContextCustomUI, queryGenerationType } = currentRoute.groupingRouteFields;
+
+  switch (currentRoute.step) {
+    case RouteStep.Mappings:
+      return (
+        <Mappings
+          onClickAddMapping={() =>
+            navigateTo(() => ({ step: RouteStep.MappingsAction, title: "Add Mapping", groupingRouteFields: {} }))
+          }
+          onClickMappingTitle={(mapping) => {
+            navigateTo(() => ({
+              step: RouteStep.Groups,
+              groupingRouteFields: { mapping },
+              title: mapping.mappingName,
+            }));
+          }}
+          onClickMappingModify={(mapping) => {
+            navigateTo(() => ({
+              step: RouteStep.MappingsAction,
+              groupingRouteFields: { mapping },
+              title: mapping.mappingName,
+            }));
+          }}
+        />);
+    case RouteStep.MappingsAction:
+      return (<MappingAction mapping={mapping} onClickCancel={goBack} onSaveSuccess={goBack} />);
+    case RouteStep.Groups:
+      if (mapping) {
+        return (
+          <Groupings
+            mapping={mapping}
+            onClickAddGroup={(qType) =>
+              navigateTo((prev) => ({
+                step: RouteStep.GroupAction,
+                groupingRouteFields: { ...prev?.groupingRouteFields, queryGenerationType: qType },
+                title: "Add Group",
+              }))
+            }
+            onClickGroupTitle={(g) =>
+              navigateTo((prev) => ({
+                step: RouteStep.Properties,
+                title: g.groupName,
+                groupingRouteFields: { ...prev?.groupingRouteFields, group: g },
+              }))
+            }
+            onClickGroupModify={(g, qType) =>
+              navigateTo((prev) => ({
+                step: RouteStep.GroupAction,
+                title: g.groupName,
+                groupingRouteFields: { ...prev?.groupingRouteFields, group: g, queryGenerationType: qType },
+              }))
+            }
+            onClickRenderContextCustomUI={(ccUI, g) =>
+              navigateTo((prev) => ({
+                step: RouteStep.GroupAction,
+                title: ccUI.displayName ?? "",
+                groupingRouteFields: { ...prev?.groupingRouteFields, group: g, groupContextCustomUI: ccUI },
+              }))
+            }
+          />
+        );
+      }
+      return null;
+    case RouteStep.GroupAction:
+      if (mapping) {
+        if (queryGenerationType) {
+          return (
+            <GroupAction
+              mappingId={mapping.id}
+              group={group}
+              onClickCancel={goBack}
+              onSaveSuccess={goBack}
```

Comment:
Are both onClickCancel and onSaveSuccess the same? Can we just use a "goBack" to replace the two?

Reply:
When I was putting the sample together, I noticed that it would be nice to separate the two events. It was hard to distinguish what actually happened. The widget itself doesn't need it as it produces the same result.